gh-140260: Fix data race in _struct module initialization with subinterpreters

ashm-dev opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-140260 -->
* Issue: gh-140260
<!-- /gh-issue-number -->

ashm-dev committed:
Fix data race in _struct module endian table initialization

ashm-dev committed:
fix doc

ashm-dev committed:
refactor(c-analyzer): add endian_tables_init_once to globals-to-fix list

ZeroIntensity commented on Modules/_struct.c:
```diff
@@ -1505,6 +1506,55 @@ static formatdef lilendian_table[] = {
+static int
+init_endian_tables(void *arg)
+{
+    (void)arg;  // Unused but required by _Py_once_fn_t signature
```
Use `Py_UNUSED(arg)` in the signature instead.

ashm-dev replied:
fixed

ashm-dev committed:
Use Py_UNUSED for unused arg in init_endian_tables

ZeroIntensity commented on Modules/_struct.c:
```diff
@@ -1505,6 +1506,55 @@ static formatdef lilendian_table[] = {
+static int
+init_endian_tables(void *arg)
+{
+    Py_UNUSED(arg);  // Unused but required by _Py_once_fn_t signature
```
I meant that you should use it in the signature, like `init_endian_tables(void *Py_UNUSED(arg))`.

ashm-dev replied:
Oops, I absentmindedly fixed it last time) Redid it properly now.

ZeroIntensity reviewed (commented):
Please add a test stressing this.

picnixz commented on Modules/_struct.c:
```diff
@@ -12,6 +12,7 @@
 #include "pycore_long.h"          // _PyLong_AsByteArray()
 #include "pycore_moduleobject.h"  // _PyModule_GetState()
 #include "pycore_weakref.h"       // FT_CLEAR_WEAKREFS()
+#include "pycore_lock.h"          // _PyOnceFlag_CallOnce()
```
Please order the includes.

ashm-dev replied:
fixed

ashm-dev committed:
fix: remove duplicate include of pycore_lock.h in _struct.c

ashm-dev committed:
add test

ZeroIntensity commented on Lib/test/test_struct.py:
```diff
@@ -800,6 +800,24 @@ def test_c_complex_round_trip(self):
+        )
+        for _ in range(3):
+            ret = support.run_in_subinterp(code)
+            self.assertEqual(ret, 0)
```
This isn't testing what we want. We need to test for initializing `_struct` in multiple interpreters concurrently, not sequentially.

ashm-dev replied:
ReDone!)

ZeroIntensity commented on Lib/test/test_struct.py:
```diff
@@ -800,6 +819,44 @@ def test_c_complex_round_trip(self):
+        # initialized concurrently.
+        # Use struct in the main interpreter first.
+        self.assertEqual(struct.unpack('>i', struct.pack('>i', 1))[0], 1)
+        self.assertEqual(struct.unpack('<i', struct.pack('<i', 1))[0], 1)
```
This is unrelated to the fix, so we don't need this in the test.

ashm-dev replied:
Agreed, deleted

ZeroIntensity commented on Lib/test/test_struct.py:
```diff
@@ -800,6 +819,44 @@ def test_c_complex_round_trip(self):
+            "x = struct.pack('>i', 1)\n"
+            "assert struct.unpack('>i', x)[0] == 1\n"
+            "y = struct.pack('<i', 1)\n"
+            "assert struct.unpack('<i', y)[0] == 1\n"
```
Same here; we only need to test `import struct` concurrently.

ashm-dev replied:
Converted to `import struct`

ZeroIntensity commented on Lib/test/test_struct.py:
```diff
@@ -800,6 +819,44 @@ def test_c_complex_round_trip(self):
+        for result in results:
+            if result.err:
+                raise result.err
+            self.assertEqual(result.ret, 0)
```
I think the whole `_Result` class is an overcomplication that we don't need here. A simple `InterpreterPoolExecutor` would be fine.

ashm-dev replied:
Understood, rewritten

ZeroIntensity commented on Modules/_struct.c:
```diff
@@ -2711,45 +2759,8 @@ _structmodule_exec(PyObject *m)
-            }
-            native++;
-        }
+    if (_PyOnceFlag_CallOnce(&endian_tables_init_once, init_endian_tables, NULL) == -1) {
```
This can never fail.

ashm-dev replied:
Rewrote it, removing if and leaving a simple call to `_PyOnceFlag_CallOnce`

picnixz replied:
In this case I would like an assert (it can never fail because the init function can never fail) or a `(void)` suppression + a comment.

ashm-dev committed:
Remove return check for endian table init

ashm-dev committed:
Use InterpreterPoolExecutor in struct tests

ZeroIntensity commented on Lib/test/test_struct.py:
```diff
@@ -1,6 +1,7 @@
 from collections import abc
 from itertools import combinations
 import array
+from concurrent.futures import InterpreterPoolExecutor
```
Please move this to the test instead.

ZeroIntensity commented on Lib/test/test_struct.py:
```diff
@@ -800,6 +801,15 @@ def test_c_complex_round_trip(self):
                     round_trip = struct.unpack(f, struct.pack(f, z))[0]
                     self.assertComplexesAreIdentical(z, round_trip)
 
+    def test_endian_table_init_subinterpreters(self):
```
We need to skip this test if subinterpreters aren't available.

ZeroIntensity commented on Lib/test/test_struct.py:
```diff
@@ -800,6 +801,15 @@ def test_c_complex_round_trip(self):
+        # concurrently in subinterpreters (gh-140260).
+        code = "import struct"
+        with InterpreterPoolExecutor(max_workers=5) as executor:
+            results = executor.map(support.run_in_subinterp, [code] * 5)
```
Don't use `support.run_in_subinterp` here, because that will create an extra subinterpreter for each subinterpreter spawned by the `InterpreterPoolExecutor`.

ZeroIntensity commented on Tools/c-analyzer/cpython/globals-to-fix.tsv:
This should be in `ignored.tsv`, not `globals-to-fix.tsv`.

ashm-dev committed:
Ignore return value of _PyOnceFlag_CallOnce

Silence compiler warning about the unused result; the init cannot fail.

ashm-dev committed:
Move endian_tables_init_once to ignored.tsv

ashm-dev committed:
Import InterpreterPoolExecutor locally in test

ashm-dev committed:
Skip subinterpreter init test unless GIL disabled

Use an exec wrapper for subinterpreter execution and assert None results
instead of calling support.run_in_subinterp

ZeroIntensity commented on Lib/test/test_struct.py:
```diff
@@ -800,6 +800,20 @@ def test_c_complex_round_trip(self):
                     round_trip = struct.unpack(f, struct.pack(f, z))[0]
                     self.assertComplexesAreIdentical(z, round_trip)
 
+    @unittest.skipUnless(support.Py_GIL_DISABLED, 'this test can only possibly fail with GIL disabled')
```
Subinterpreters are supported on the GIL-ful build too. I meant that we should skip this test (via `raise unittest.SkipTest`) when importing `InterpreterPoolExecutor` fails.

ZeroIntensity commented on Lib/test/test_struct.py:
```diff
@@ -800,6 +800,20 @@ def test_c_complex_round_trip(self):
+
+        code = "import struct"
+        with InterpreterPoolExecutor(max_workers=5) as executor:
+            results = executor.map(_exec_code, [code] * 5)
```
You should be able to just use `exec` here.

ashm-dev committed:
Inline exec and wrap skipUnless args

ashm-dev committed:
Skip test when InterpreterPoolExecutor missing

ashm-dev committed:
Fix executor.map usage in test_struct

ashm-dev committed:
Pass exec to executor.map

ZeroIntensity committed:
Merge branch 'main' into tsan

ZeroIntensity reviewed (approved):
LGTM

kumaraditya303 committed:
Merge branch 'main' into tsan

ashm-dev committed:
fix: tests

kumaraditya303 committed:
Merge branch 'main' into tsan
